From 0ae57dfc079926b6516017c5ee20eb4d267a3b06 Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Fri, 22 May 2026 14:17:08 +0100
Subject: [PATCH] CEC Support

Add support for the special new features in pulseaduio-qt for CEC
support
---
 applet/ListItemBase.qml            | 34 ++++++++++++++++++++
 src/kcm/ui/DeviceListItem.qml      |  1 +
 src/kcm/ui/StreamListItem.qml      |  2 +-
 src/kded/audioshortcutsservice.cpp | 51 +++++++++++++++++++++++-------
 4 files changed, 75 insertions(+), 13 deletions(-)

diff --git a/applet/ListItemBase.qml b/applet/ListItemBase.qml
index 0e0effe2..15a9f072 100644
--- a/applet/ListItemBase.qml
+++ b/applet/ListItemBase.qml
@@ -214,8 +214,42 @@ PC3.ItemDelegate {
                 }
             }
 
+            RowLayout {
+                visible: !item.model.VolumeWritable
+                uniformCellSizes: true
+                PC3.Button {
+                    Layout.fillWidth: true
+                    text: i18n("Volume Down")
+                    icon.name: "audio-volume-low-symbolic"
+                    icon.width: Kirigami.Units.iconSizes.medium
+                    icon.height: Kirigami.Units.iconSizes.medium
+                    onClicked: item.model.PulseObject.volumeDown();
+                    display: PC3.AbstractButton.TextUnderIcon
+
+                }
+                PC3.Button {
+                    Layout.fillWidth: true
+                    text: i18n("Volume Up")
+                    display: PC3.AbstractButton.TextUnderIcon
+                    icon.width: Kirigami.Units.iconSizes.medium
+                    icon.height: Kirigami.Units.iconSizes.medium
+                    icon.name: "audio-volume-high-symbolic"
+                    onClicked: item.model.PulseObject.volumeUp();
+                }
+                PC3.Button {
+                    Layout.fillWidth: true
+                    text: i18n("Toggle Mute")
+                    display: PC3.AbstractButton.TextUnderIcon
+                    icon.width: Kirigami.Units.iconSizes.medium
+                    icon.height: Kirigami.Units.iconSizes.medium
+                    icon.name: "audio-volume-muted"
+                    onClicked: item.model.PulseObject.toggleMute();
+                }
+            }
+
             RowLayout {
                 spacing: Kirigami.Units.smallSpacing
+                visible: item.model.VolumeWritable
 
                 SmallToolButton {
                     id: muteButton
diff --git a/src/kcm/ui/DeviceListItem.qml b/src/kcm/ui/DeviceListItem.qml
index 937d478e..15c453f8 100644
--- a/src/kcm/ui/DeviceListItem.qml
+++ b/src/kcm/ui/DeviceListItem.qml
@@ -184,6 +184,7 @@ ColumnLayout {
             muted: Muted
             onToggled: Muted = checked
             toolTipText: !currentPort ? Description : currentPort.description
+            enabled: delegate.pulseObject.muteWritable
         }
 
         GridLayout {
diff --git a/src/kcm/ui/StreamListItem.qml b/src/kcm/ui/StreamListItem.qml
index 2aa655f0..807ee3dd 100644
--- a/src/kcm/ui/StreamListItem.qml
+++ b/src/kcm/ui/StreamListItem.qml
@@ -89,6 +89,7 @@ ColumnLayout {
             }
 
             RowLayout {
+                visible: HasVolume
                 MuteButton {
                     Layout.alignment: Qt.AlignTop
                     muted: Muted
@@ -99,7 +100,6 @@ ColumnLayout {
                 VolumeSlider {
                     id: volumeSlider
                     Layout.alignment: Qt.AlignVCenter
-                    visible:  HasVolume
                 }
             }
         }
diff --git a/src/kded/audioshortcutsservice.cpp b/src/kded/audioshortcutsservice.cpp
index 19913a5a..f389b974 100644
--- a/src/kded/audioshortcutsservice.cpp
+++ b/src/kded/audioshortcutsservice.cpp
@@ -88,9 +88,14 @@ AudioShortcutsService::AudioShortcutsService(QObject *parent, const QList<QVaria
         if (!sink) {
             return;
         }
-        int percent = changeVolumePercent(sink, m_globalConfig->volumeStep());
-        showVolume(percent);
-        playFeedback(-1);
+        if (sink->isVolumeWritable()) {
+            int percent = changeVolumePercent(sink, m_globalConfig->volumeStep());
+            showVolume(percent);
+            playFeedback(-1);
+        } else {
+            m_osdDBusInterface->showText("audio-volume-high", i18n("Volume Up"));
+            sink->volumeUp();
+        }
     });
 
     QAction *volumeDownAction = new QAction(this);
@@ -103,9 +108,14 @@ AudioShortcutsService::AudioShortcutsService(QObject *parent, const QList<QVaria
         if (!sink) {
             return;
         }
-        int percent = changeVolumePercent(sink, -m_globalConfig->volumeStep());
-        showVolume(percent);
-        playFeedback(-1);
+        if (sink->isVolumeWritable()) {
+            int percent = changeVolumePercent(sink, -m_globalConfig->volumeStep());
+            showVolume(percent);
+            playFeedback(-1);
+        } else {
+            m_osdDBusInterface->showText("audio-volume-low", i18n("Volume Down"));
+            sink->volumeDown();
+        }
     });
 
     QAction *volumeUpSmallAction = new QAction(this);
@@ -118,9 +128,14 @@ AudioShortcutsService::AudioShortcutsService(QObject *parent, const QList<QVaria
         if (!sink) {
             return;
         }
-        int percent = changeVolumePercent(sink, 1);
-        showVolume(percent);
-        playFeedback(-1);
+        if (sink->isVolumeWritable()) {
+            int percent = changeVolumePercent(sink, 1);
+            showVolume(percent);
+            playFeedback(-1);
+        } else {
+            m_osdDBusInterface->showText("audio-volume-high", i18n("Volume Up"));
+            sink->volumeDown();
+        }
     });
 
     QAction *volumeDownSmallAction = new QAction(this);
@@ -133,9 +148,14 @@ AudioShortcutsService::AudioShortcutsService(QObject *parent, const QList<QVaria
         if (!sink) {
             return;
         }
-        int percent = changeVolumePercent(sink, -1);
-        showVolume(percent);
-        playFeedback(-1);
+        if (sink->isVolumeWritable()) {
+            int percent = changeVolumePercent(sink, -1);
+            showVolume(percent);
+            playFeedback(-1);
+        } else {
+            m_osdDBusInterface->showText("audio-volume-low", i18n("Volume Down"));
+            sink->volumeDown();
+        }
     });
 
     QAction *volumeUpMicAction = new QAction(this);
@@ -172,6 +192,13 @@ AudioShortcutsService::AudioShortcutsService(QObject *parent, const QList<QVaria
     muteAction->setText(i18n("Mute"));
     muteAction->setShortcut(Qt::Key_VolumeMute);
     connect(muteAction, &QAction::triggered, this, [this]() {
+        auto sink = m_preferredDevice.sink();
+        if (sink && !sink->isMuteWritable()) {
+            m_osdDBusInterface->showText("audio-volume-muted", i18n("Toggle Mute"));
+            sink->toggleMute();
+            return;
+        }
+
         if (m_globalConfig->globalMuteSinks()) {
             disableGlobalSinkMute();
         } else {
-- 
2.55.0

